### ISSUE TITLE
[Question] Does Pyro have anything comparable to Craft/Yii's behaviors?

### ISSUE THREAD

**Issue by mattcdavis1** [pyrocms/pyrocms#3722]
One thing that i've found very useful when working with Craft / Yii is the "Behaviors" system that allows associating a class / object with an existing model without having to extend it.  The way i typically utilize this is via the onPopulateModel event in craft.  

This could be really useful with Pyro since the streams module puts the models in the storage directory making it a bit awkward to try and extend them.  Instead, if there was a "PopulateModel" event that fired and allowed me to associate my additional methods with the existing (auto-generate) model that would be great. 

Typically i'll handle this based on naming convention so if the stream name is "Photos", in the event handler for onPopulateModel i'll check if a class exists "PhotosBehavior" and associate that behavior with the model.  Then, any methods in PhotosBehavior are now available on the model as well.

Or even with the more traditional extending approach as long as there is a way to easily add additional methods to auto generated models that would be great.  Like maybe when Pyro is generating the storage models it either looks for a class to extend based on naming convention or possibly injects a trait based on naming convention (although traits are more limited since we can't override methods and call parent methods).

**Comment by RyanThompson** [comment]
This is called a "hook": http://pyrocms.com/documentation/streams-platform/hooks

Models and presenters among some others are hookable out of the box.

**Comment by mattcdavis1** [comment]
nice!  Would be helpful to see an example of registering / using a hook with a stream generated model.  Also, does the callback function get an instance of the model?

**Comment by RyanThompson** [comment]
Not unless:

1.) The handler is a method on the instance - then it’s not needed.
2.) The handler is a closure that can use ($entry) the model - then it’s not needed.
3.) The handler is passed the instance in it’s payload during fire($trigger, $payload);

**Comment by mattcdavis1** [comment]
hmm - hard to follow what you're saying without seeing in context.  I'll have to play around with this a bit and see how that goes.

**Comment by RyanThompson** [comment]
Just that callbacks only get the data available to them when it's fired. When you call fire() method you tell it what callback trigger and what payload to send to the callback. But by nature if your callback is handled by a method ON that class. Of course you have access to it. That kinda thing: https://github.com/anomalylabs/streams-platform/blob/master/src/Traits/FiresCallbacks.php#L75
